graylog2-server/src/main/java/org/graylog2/streams/StreamRouterEngine.java

```diff
@@ -83,33 +75,49 @@ public StreamRouterEngine(@Assisted List<Stream> streams,
         this.streamProcessingTimeout = streamFaultManager.getStreamProcessingTimeout();
         this.fingerprint = new StreamListFingerprint(streams).getFingerprint();
 
-        for (final Stream stream : streams) {
-            for (final StreamRule streamRule : stream.getStreamRules()) {
+        final List<Rule> presenceRules = Lists.newArrayList();
+        final List<Rule> exactRules = Lists.newArrayList();
+        final List<Rule> greaterRules = Lists.newArrayList();
+        final List<Rule> smallerRules = Lists.newArrayList();
+        final List<Rule> regexRules = Lists.newArrayList();
+
+        for (Stream stream : streams) {
+            final boolean sufficient = stream.getMatchingType() == Stream.MatchingType.OR;
+            for (StreamRule streamRule : stream.getStreamRules()) {
+                final Rule rule;
                 try {
-                    final Rule rule = new Rule(stream, streamRule);
-
-                    switch (streamRule.getType()) {
-                        case EXACT:
-                            addRule(exactRules, exactFields, streamRule.getField(), rule);
-                            break;
-                        case GREATER:
-                            addRule(greaterRules, greaterFields, streamRule.getField(), rule);
-                            break;
-                        case SMALLER:
-                            addRule(smallerRules, smallerFields, streamRule.getField(), rule);
-                            break;
-                        case REGEX:
-                            addRule(regexRules, regexFields, streamRule.getField(), rule);
-                            break;
-                        case PRESENCE:
-                            addRule(presenceRules, presenceFields, streamRule.getField(), rule);
-                            break;
-                    }
+                    rule = new Rule(stream, streamRule, sufficient);
                 } catch (InvalidStreamRuleTypeException e) {
                     LOG.warn("Invalid stream rule type. Skipping matching for this rule. " + e.getMessage(), e);
+                    continue;
+                }
+                switch (streamRule.getType()) {
+                    case PRESENCE:
+                        presenceRules.add(rule);
+                        break;
+                    case EXACT:
+                        exactRules.add(rule);
+                        break;
+                    case GREATER:
+                        greaterRules.add(rule);
+                        break;
+                    case SMALLER:
+                        smallerRules.add(rule);
+                        break;
+                    case REGEX:
+                        regexRules.add(rule);
+                        break;
                 }
             }
         }
+
+        this.rulesList = new ImmutableList.Builder<Rule>()
```

Comment:
Having looked at the surrounding code, this class is initialized every second (even if the rules haven't been changed) and `rulesList` is only used internally in this class, so I think initializing the list with the correct size would yield some improvements and is worth the disadvantage(?) over having a mutable collection instead of an immutable one.